### DISCUSSION TITLE
Link to headings and blocks

### DISCUSSION THREAD

**Discussion by xygn** [srid/emanote#105] [Ideas]
Hi!

I think linking up headings and blocks would be cool feature as you can directly reference a certain heading of a note rather than the entire note itself. Would be very functional for longer notes.

Additionally, in reference to #11 a small hamburger/dropdown can show the outline of the document, similar to how GitHub does it:
![image](https://user-images.githubusercontent.com/63365609/124807048-fdae6100-df7a-11eb-8eb6-5aefc9bcfc1e.png)

**Same-file heading link:**
`[[#1 Venous Thrombophlebitis]]`

**Cross-file heading link:**
`[[4. Immunity#Proteinaceous Antigens]]`

**Same-file block link:**
`[[#^79b23f]]` and the presence of `^79b23f` in the block linked

**Cross-file block link:**
`[[1. Cell Injury#^565948]]` and the presence of `^565948` in the block linked inside the original file

**From Obsidian Docs,**
[Headings](https://help.obsidian.md/How+to/Internal+link#Link+to+headings)
[Blocks](https://help.obsidian.md/How+to/Link+to+blocks)

Obsidian's block linking is _non-standard_ and so far only Obsidian seems to support it. Maybe the heading feature can be prioritized over blocks, but both of them can be useful in their own way

> Please be aware that block reference is not part of standard Markdown, but rather Obsidian's specific flavor of Markdown. That means these links will not work outside of Obsidian.
> 
> You can still find referenced block by searching for the block ID, although that's a rather manual process. The connection is not lost as long as you have software that can search in a folder.

**Comment by srid** [comment]
> [[1. Cell Injury#^565948]] and the presence of ^565948 in the block linked inside the original file

What's the syntax used in `Cell Injury.md` to define the block 565948? As in, what does the actual Markdown look like?

I'm not worried about the block *linking* syntax itself (it looks reasonable enough to be adopted by others), but for block *definition* - I wonder if Obsidian is relying on some non-standard syntax here (which would be bad). The 'standard' syntax (at least as far as commonmark is concerned) is to use [fenced_divs](https://github.com/jgm/commonmark-hs/blob/master/commonmark-extensions/test/fenced_divs.md) with an ID, for instance:

```markdown
Cell Injury is ...

:::{#565948}
- Foo
- Bar
:::

More content
```

**Reply by xygn** [reply]
> if Obsidian is relying on some non-standard syntax here

The syntax used for block definition in Cell Injury.md is non-standard as expected:

```
Content

- Genes
	1. BCL-2 ^565948
	2. BCL-XL
	3. MCL-1 gene 

More content
```

It leaves `^565948` beside the block to be linked. Here it is the list item `1. BCL-2`

Here's an implementation of it in their website : https://help.obsidian.md/How+to/Link+to+blocks#^dcf64c with `^dcf64c` being the block ID linked

The block ID `^alphanumeric` is hidden on the webpage and in the preview mode (of Obsidian)

The block ID can be usergenerated too, as long as it is unique and follows this:

> Make sure there's at least one space if you're putting the ID at end of a line.
> Note that only letters, numbers, and dashes are allowed in manual block IDs.

**Comment by tomtomjhj** [comment]
To support link to heading, we first need to choose the rule for generating the heading's identifier. There are several flavors for this.
* Obsidian: Preserve case, shrink spaces, and remove some special characters. For example, the heading ``# BIG heading   `code` ~+-!@#`` can be linked as `[[file#BIG heading code -]]`.
* Pandoc: <https://pandoc.org/MANUAL.html#extension-auto_identifiers>. It also supports manually assigning the id <https://pandoc.org/MANUAL.html#extension-header_attributes>.
* GFM: https://pandoc.org/MANUAL.html#extension-gfm_auto_identifiers

**Comment by srid** [comment]
Heading links should work in https://github.com/srid/emanote/pull/268

**Comment by srid** [comment]
Larger feature: **heading-awareness** https://github.com/srid/emanote/issues/272

**Comment by ChenghaoMou** [comment]
Happy New Year!

I have made block-level embedding somewhat usable with a [Python preprocessing script](https://github.com/ChenghaoMou/chenghaomou.github.io/blob/main/export/block.py) by simply replacing the block wikilink with the referenced block. See [example](https://chenghaomou.github.io/notes/20220709195323). Unfortunately, the linkage will be lost since it is done before emanote's processing.

The block identifier could appear inline at the end of a block or after a block (code, title, quote etc.). @srid Please let me know if you are interested in me adding this feature to emanote. However, I am very much a newbie to Haskell or JS, and pointers would be much appreciated.

**Reply by srid** [reply]
@ChenghaoMou I'm travelling right now, hence slow to respond. But yes I'd be happy for you to read this feature to Emanote. Once I get a chance I'll give some initial points for you. Meanwhile, you can also [chat in Zulip](https://nixos.zulipchat.com/#narrow/stream/416826-website).
